onset_maproom/agronomy.py

```diff
@@ -0,0 +1,535 @@
+    sminit0 = xr.DataArray(coords=dict({time_coord: [t0.values]}), data=[sminit])
+    # Initialize Ks
+    ks = 1
+    if rho is not None:
+        raw = calibrate_available_water(taw, rho)
+        ks = single_stress_coeff(
+            sminit0,
+            taw,
+            raw.isel({time_coord: 0}, missing_dims="ignore"),
+            time_coord=time_coord,
+        ).squeeze(time_coord)
+    # Create or Initialize Kc
+    if kc_params is None:
+        kc = xr.ones_like(water_balance.et)
+    else:
+        if p_d is not None:
+            kc = kc_interpolation(p_d, kc_params, time_coord=time_coord)
+        else:
+            p_d_find = planting_date(
+                sminit0, sm_threshold, time_coord=time_coord
+            ).expand_dims(dim=time_coord)
+            kc = kc_interpolation(p_d_find, kc_params, time_coord=time_coord)
+    # Initializaing sm, drainage, et_crop, et_crop_red
+    if time_coord in kc.dims:
+        kc0 = kc.sel({time_coord: water_balance.soil_moisture[time_coord][0]})
+    else:
+        kc0 = kc
+    water_balance.et_crop[{time_coord: 0}] = crop_evapotranspiration(
+        water_balance.et.isel({time_coord: 0}).expand_dims(dim=time_coord),
+        kc0,
+        time_coord=time_coord,
+    ).squeeze(time_coord)
+    water_balance.et_crop_red[{time_coord: 0}] = reduce_crop_evapotranspiration(
+        water_balance.et_crop
+        .isel({time_coord: 0}).expand_dims(dim=time_coord),
+        ks,
+    ).squeeze(time_coord)
+    water_balance.soil_moisture[{time_coord: 0}] = (
+        sminit
+        + water_balance.peffective.isel({time_coord: 0})
+        - water_balance.et_crop_red.isel({time_coord: 0})
+    ).clip(min=0)
+    water_balance.drainage[{time_coord: 0}] = (
+        water_balance.soil_moisture.isel({time_coord: 0})
+        - taw
+    ).clip(min=0)
+    water_balance.soil_moisture[{time_coord: 0}] = (
+        water_balance.soil_moisture.isel({time_coord: 0})
+        - water_balance.drainage.isel({time_coord: 0})
+    )
+    # Looping on time_coord
+    # Get time_coord info
+    time_coord_size = water_balance.peffective[time_coord].size
+    for i in range(1, time_coord_size):
+        if rho is not None:
+            ks = single_stress_coeff(
+                water_balance.soil_moisture
+                .isel({time_coord: i - 1}).expand_dims(dim=time_coord),
+                taw,
+                raw,
+                time_coord=time_coord,
+            ).squeeze(time_coord)
+        if kc_params is not None and p_d is None:
+            p_d_iter = planting_date(
+                water_balance.soil_moisture
+                .isel({time_coord: i - 1}).expand_dims(dim=time_coord),
+                sm_threshold,
+                time_coord=time_coord,
+            )
+            # it's desired and useful to update Kc if
+            # p_d was NaT and is now found
+            kc_updates = np.isnat(p_d_find) and not np.isnat(p_d_iter)
```

Comment:
While investigating the performance issue, I happened to print the value of `kc_updates` here and was surprised to find that it's a Boolean scalar. I think you intended for it to be an array of Booleans, since `p_d_find` and `p_d_iter` are both arrays. The problem is that `and` is a normal python operator, not an array operator.

The solution to the immediate problem is to use `&` instead of `and`. But the fact that this bug went unnoticed raises bigger issues. First of all, why do the tests pass despite this bug? Can we add a test case that would have caught it? Second, why are `p_d_find` and `p_d_iter` arrays anyway? Is that so we can handle multiple years? But we're not handling multiple years yet. Can we simplify this for now now by making them scalars instead?

Reply:
whoops, and the `not` should be replaced with `~`.

Reply:
I tried making the changes I suggested here and below (replacing `and` with `&`, `not` with `~`), and it doesn't work. I may have misunderstood the intent. Unless you know how to fix it, maybe we should work through this together.

Reply:
Did you notice I was terrible with Python comparisons? :)
Let me have a look for myself. In the meantime, one piece of information for you: p_d/_find/_iter can vary like soil_moisture if they are estimated inside the function, so space at least. This is probably an initialization problem because initial soil moisture is sminit that has been a constant/scalar in all the tests. In the code at some point I bothered to give it a time dimension, I guess I should broadcast it against soil_moisture dims but T. And cover the cases where sminit is scalar or xarray (because here again, someone could have an sminit that depends on space already, or just want to test several initial conditions with some made up dim...)

Maybe I should just concat sminit to soil_moisture and keep going with soil_moisture... That's going to complicate the positioning selection mentioned earlier. I'll give it a thought, if you have ideas, I'll take them.

Reply:
I wonder now if that explains the shift by 1 day of the results of the tests mentioned earlier...

Reply:
I am going to mark this resolved but also add note to think of more tests. I did a bunch of prints and the behavior expected with respect to planting date, et and et_crop seemed appropriate. It may mean that it was wrong before. But I figure I would make best use of my time not to figure that out, but rather to think of more tests